### PR TITLE
[13.0][MIG] purchase

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Migration of `purchase`.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/purchase/migrations/13.0.1.2/post-migration.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        openupgrade.logged_query(
+            cr, """
+            UPDATE purchase_order_line pol
+            SET qty_received_method = 'stock_moves'
+            FROM product_product pp
+            LEFT JOIN product_template pt ON pp.product_tmpl_id = pt.id
+            WHERE pol.product_id = pp.id AND pol.display_type IS NULL
+                AND pt.type IN ('consu', 'product')
+            """
+        )
```
Comment (kos94ok-3D) [inline addons/purchase/migrations/13.0.1.2/post-migration.py]: <pre>SET qty_received_method = 'stock_moves', qty_received_manual = 0.0</pre>

**Review by kos94ok-3D — COMMENTED** [review]
I think need to keep a behave written in the `_inverse_qty_received` function
изображение

File: addons/purchase/migrations/13.0.1.2/post-migration.py (hunk 1 above)
Comment (MiquelRForgeFlow) [reply addons/purchase/migrations/13.0.1.2/post-migration.py]: I set it to NULL instead, ok?

**Review by kos94ok-3D — APPROVED** [review]

File: addons/purchase/migrations/13.0.1.2/post-migration.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+
+def move_fields_from_invoice_to_moves(env):
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE account_move_line aml
+        SET purchase_line_id = ail.purchase_line_id
+        FROM account_invoice_line ail
+        WHERE aml.old_invoice_line_id = ail.id"""
+    )
```
Comment (legalsylvain) [inline addons/purchase/migrations/13.0.1.2/post-migration.py:14]: I think adding ``and ail.purchase_line_id is not null`` could make the request faster, avoiding false in many cases.

File: addons/purchase/migrations/13.0.1.2/post-migration.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        JOIN account_move am ON am.old_invoice_id = ai.id
+        WHERE ampor.{} = ai.id
+        """.format(openupgrade.get_legacy_name('account_invoice_id'))
+    )
+
+
+def fill_purchase_order_line_qty_received_method(cr):
+    if openupgrade.column_exists(cr, 'purchase_order', 'picking_type_id'):
+        # purchase_stock is installed
+        # set qty_delivered_method = 'stock_moves'
```
Comment (legalsylvain) [inline addons/purchase/migrations/13.0.1.2/post-migration.py]: There is a function to test if a module is installed.

**Review by legalsylvain — COMMENTED** [review]
Some minor remarks. Lgtm otherwise. Thanks !

File: addons/purchase/migrations/13.0.1.2/post-migration.py (hunk 3 above)
Comment (MiquelRForgeFlow) [reply addons/purchase/migrations/13.0.1.2/post-migration.py]: Is it better that function than using that column_exists?

Comment (kos94ok-3D) [reply addons/purchase/migrations/13.0.1.2/post-migration.py]: It is not always possible, because the module can be merged into another

**Review by Abranes — APPROVED** [review]

File: addons/purchase/migrations/13.0.1.2/pre-migration.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+def migrate(env, version):
+    openupgrade.copy_columns(env.cr, _column_copies)
+    openupgrade.rename_columns(env.cr, _column_renames)
+    openupgrade.rename_tables(env.cr, _table_renames)
+    openupgrade.rename_xmlids(env.cr, _xmlid_renames)
+    openupgrade.logged_query(
+        env.cr, """
+        ALTER TABLE purchase_order_line
+        ADD COLUMN qty_received_method varchar""",
+    )
```
Comment (ThomasBinsfeld) [inline addons/purchase/migrations/13.0.1.2/pre-migration.py:34]: `rename_columns` and `rename_tables` preserve the `NOT NULL` on the legacy `account_invoice_id` column. That can cause some problems. I think the `NOT NULL` should be dropped here.

```suggestion
    )
    openupgrade.logged_query(
        env.cr, """
        ALTER TABLE account_move_purchase_order_rel
        ALTER COLUMN openupgrade_legacy_13_0_account_invoice_id
        DROP NOT NULL""",
    )
```

Comment (MiquelRForgeFlow) [reply addons/purchase/migrations/13.0.1.2/pre-migration.py:34]: Now rename is avoided.

**Comment by bishalgit** [comment]
@MiquelRForgeFlow  Should I review this one? Why is this still not merged?

**Comment by abdounasser202** [comment]
I think the migration of this module blocks the one from sale_purchase

**Comment by MiquelRForgeFlow** [comment]
> Should I review this one?

Reviews are always welcomed.

> Why is this still not merged?

Because @pedrobaeza still has not reviewed it.

**Comment by pedrobaeza** [comment]
Well, this one is heavy and some experienced reviewer is a good choice for not merging something not completed. Anyway, right now the CI is red, so it must be solved before merging. I will review this meanwhile. @MiquelRForgeFlow please check.

**Comment by MiquelRForgeFlow** [comment]
@pedrobaeza The error in CI is due to purchase_stock, which should not be taken into account, as it will follow this PR.

**Comment by kos94ok-3D** [comment]
@MiquelRForgeFlow  Runbot is red so maybe better move code from https://github.com/OCA/OpenUpgrade/pull/2314 here.
@pedrobaeza Do you plan to review it until summer?

**Comment by pedrobaeza** [comment]
I have a local branch with the needed changes.

**Review by pedrobaeza — APPROVED** [review]
I have reviewed and performed latest fine-tune

**Pull request merged**